### Fake news Detection

Fake news is a type of yellow journalism that incorporates pieces of news that may be hoaxes and is widely disseminated via social media and other internet platforms. This is frequently done to promote or impose specific views, and it is frequently accomplished through political agendas. Such news pieces may contain misleading and/or inflated claims, and algorithms may spread them, trapping users in a filter bubble.

In [1]:
# Importing library

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

import warnings

In [3]:
news = pd.read_csv(r'C:\Users\kamlesh\Downloads\Fake_news\news.csv')

In [4]:
news.head(5)

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [5]:
news.shape

(6335, 4)

In [6]:
news.head(4)

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE

In [7]:
news.isnull().sum()

Unnamed: 0    0
title         0
text          0
label         0
dtype: int64

In [8]:
news.duplicated().sum()

0

In [9]:
news.dtypes

Unnamed: 0     int64
title         object
text          object
label         object
dtype: object

* No Null Vales & No Duplicated rows

In [10]:
labels = news.label

In [11]:
import itertools
from sklearn.model_selection import train_test_split

* Split dataset into training and testing sets

In [12]:
x_train,x_test,y_train,y_test=train_test_split(news['text'], labels, test_size=0.2, random_state=7)

* Let’s initialize a TfidfVectorizer with stop words from the English language and a maximum document frequency of 0.7 

In [13]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [14]:
# Initialize - Tfidfvectorizer
vector = TfidfVectorizer(stop_words='english',max_df= 0.7)

In [15]:
#fit and tramsform trainig and test set

In [16]:
tf_train=vector.fit_transform(x_train) 
tf_test=vector.transform(x_test)

* we’ll initialize a PassiveAggressiveClassifier. This is. We’ll fit this on tfidf_train and y_train.

In [17]:
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

In [18]:
pac=PassiveAggressiveClassifier(max_iter=50)
pac.fit(tf_train,y_train)

#Predict on the test set and calculate accuracy
y_pred=pac.predict(tf_test)
score=accuracy_score(y_test,y_pred)
print(f'Accuracy: {round(score*100,2)}%')

Accuracy: 92.58%


we got 92.58% of accuracy with this model, so lets print out the confusion matrix to gain insights into the number of false and true

In [19]:
confusion_matrix (y_test,y_pred, labels=['FAKE','REAL'])

array([[590,  48],
       [ 46, 583]], dtype=int64)

* we have 590 true positives, 583 true negatives, 46 false positives, and 48 false negatives.